### PR TITLE
Add Fine-Tuning for Wav2Vec2

### LINKED ISSUE #10231: Wav2Vec2 finetune
# 🚀 Feature request

Can you please share code to how to finetune transformers.Wav2Vec2ForCTC, or maybe on how to give labels to the model in order to get loss?

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds the possibility to finetune Wav2Vec2 on a downstream task. I ran a couple of experiments and I think the training is pretty stable now, see this training run *e.g.*:

https://wandb.ai/patrickvonplaten/huggingface/reports/Project-Dashboard--Vmlldzo0OTI0OTc?accessToken=8azw8iyxnbiqd4ytxcgm4hbnfh3x1b2c9l2eyfqfzdqw7l0icreljc9qpx0rkl6f

Once this is merged, I will make a nice forum post and link 1,2 notebooks.

## Who can review?

Would be great if @sgugger @LysandreJik and @patil-suraj could review.

### CODE REVIEW

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -726,6 +869,36 @@ def forward(
 
         hidden_states = self.feature_projection(hidden_states)
 
+        if self.config.apply_spec_augment and self.training:
```
Comment (SeanNaren) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:872]: I'm curious, can we apply spec augment to this model? It's trained on the raw waveform rather than spectrograms, so is this more a conv extracted feature drop out?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -893,8 +1066,38 @@ def forward(
...
+
+            with torch.backends.cudnn.flags(enabled=False):
+                loss = F.ctc_loss(
+                    log_probs,
+                    flattened_targets,
+                    input_lengths,
+                    target_lengths,
+                    blank=self.config.pad_token_id,
+                    reduction="sum",
+                    zero_infinity=False,
```
Comment (SeanNaren) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1094]: Ah! I would expose this parameter to the user if possible or set this to True by default, there are a few cases where CTC can throw a infinite loss at scale, and you probably want to ignore it

**Comment by SeanNaren** [comment]
This is really nice, and the piece that will make the wav2vec 2.0 stuff awesome and more readily available! let me know if I can assist in testing/whatnot :)

File: src/transformers/tokenization_utils_base.py [hunk 3]
```diff
@@ -2573,6 +2573,7 @@ def pad(
         padding: Union[bool, str, PaddingStrategy] = True,
         max_length: Optional[int] = None,
         pad_to_multiple_of: Optional[int] = None,
+        pad_token_id: Optional[int] = None,
```
Comment (patrickvonplaten) [inline src/transformers/tokenization_utils_base.py:2576]: not sure whether this is the way to go

**Comment by flozi00** [comment]
Pretty cool @patrickvonplaten 
Only have two questions:
What about self training for low ressource languages ?
Is it deepspeed compatible ?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:822]: exactly! spec_augment is applied to the features extracted from the conv layers

File: examples/research_projects/wav2vec2/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+## Fine-tuning Wav2Vec2
+
+The `run_training.py` script allows one to fine-tune a one of the pretrained Wav2Vec2 model that can be found [here](https://huggingface.co/models?search=facebook/wav2vec2).
+
+This finetuning script can also be run as a google colab [TODO: here]( ).
```
Comment (patrickvonplaten) [inline examples/research_projects/wav2vec2/README.md:5]: I'll make the notebook once I have completed all the fine-tuning runs.

File: examples/research_projects/wav2vec2/run_training.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/env python3
```
Comment (patrickvonplaten) [inline examples/research_projects/wav2vec2/run_training.py:1]: @LysandreJik, @sgugger 

I think it makes sense here to first put everything under `research_projects` and once it's heavily used we can maybe move the collator to `src/transformers` and move this script under `examples/speech`

File: src/transformers/models/wav2vec2/configuration_wav2vec2.py [hunk 6]
```diff
@@ -72,7 +72,7 @@ class Wav2Vec2Config(PretrainedConfig):
             The non-linear activation function (function or string) in the 1D convolutional layers of the feature
             extractor. If string, :obj:`"gelu"`, :obj:`"relu"`, :obj:`"selu"` and :obj:`"gelu_new"` are supported.
         conv_dim (:obj:`Tuple[int]`, `optional`, defaults to :obj:`(512, 512, 512, 512, 512, 512, 512)`):
-            A tuple of integers defining the number of input and output channels of each 1D convolutional layer in the
+            A tuple of integers defining the number of input and output features of each 1D convolutional layer in the
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/configuration_wav2vec2.py:75]:
```suggestion
            A tuple of integers defining the number of input and output channels of each 1D convolutional layer in the
```

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 7]
```diff
@@ -44,6 +44,77 @@
 ]
 
 
+def _compute_mask_indices(
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:47]: This can be done in a better/faster way when coding everything in pytorch -> but training works well as it is for the moment. Will put it up as a "First Good Issue" to convert this function fully into PyTorch

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 8]
```diff
@@ -893,8 +1049,38 @@ def forward(
...
+
+            # retrieve loss input_lengths from attention_mask
+            attention_mask = (
+                attention_mask if attention_mask is not None else torch.ones_like(input_values, dtype=torch.long)
+            )
+            input_lengths = self._get_feat_extract_output_lengths(attention_mask.sum(-1))
+
+            # assuming that padded tokens are filled with -100
+            # when not being attended to
+            labels_mask = labels >= 0
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1063]: That's important we assume that padded labels have been replaced by -100 before training or else the fine-tuning would crash. The second option would be to actually check whether the labels contain a `pad_token_id` and to automatically replace this token then. However I think the first option is more in line with how we currently do fine-tuning.

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -509,11 +509,18 @@ def _get_train_sampler(self) -> Optional[torch.utils.data.sampler.Sampler]:
 
         # Build the sampler.
         if self.args.group_by_length:
+            model_input_name = self.tokenizer.model_input_names[0] if self.tokenizer is not None else None
```
Comment (patrickvonplaten) [inline src/transformers/trainer.py:512]: @sgugger - the `LengthGroupedSampler` is very important to speed-up training for Wav2Vec2, but the processor has `input_values` as its model input name so that we have to pass this to the `LengthGroupedSampler`. See changes in `src/transformers/trainer_pt_utils.py`

File: examples/research_projects/wav2vec2/run_training.py (hunk 5 above)
Comment (LysandreJik) [reply examples/research_projects/wav2vec2/run_training.py:1]: Yes, this works for me.

File: examples/research_projects/wav2vec2/run_training.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+from typing import Any, Dict, List, Optional, Union
+
+import datasets
+import numpy as np
+import torch
+import torch.nn as nn
+from packaging import version
+
+import soundfile as sf
+from jiwer import wer
```
Comment (LysandreJik) [inline examples/research_projects/wav2vec2/run_training.py:12]: Should we have a `requirements.txt` in that research project?

File: src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to_pytorch.py [hunk 11]
```diff
@@ -20,34 +20,39 @@
...
-    "self_attn.out_proj": "wav2vec2.encoder.layers.*.attention.out_proj",
-    "self_attn_layer_norm": "wav2vec2.encoder.layers.*.layer_norm",
-    "fc1": "wav2vec2.encoder.layers.*.feed_forward.intermediate_dense",
-    "fc2": "wav2vec2.encoder.layers.*.feed_forward.output_dense",
-    "final_layer_norm": "wav2vec2.encoder.layers.*.final_layer_norm",
-    "encoder.layer_norm": "wav2vec2.encoder.layer_norm",
-    "w2v_model.layer_norm": "wav2vec2.feature_projection.layer_norm",
+    "post_extract_proj": "feature_projection.projection",
+    "encoder.pos_conv.0": "encoder.pos_conv_embed.conv",
+    "self_attn.k_proj": "encoder.layers.*.attention.k_proj",
```
Comment (LysandreJik) [inline src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to_pytorch.py:32]: Will the weights on the hub need to be updated?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 7 above)
Comment (LysandreJik) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:47]: Given that it is used in the forward method and that it implies a numpy <> torch conversion, having that be all in PyTorch would probably result in a big performance boost, wouldn't it?

Ok to put it up as a good first issue!

File: examples/research_projects/wav2vec2/README.md [hunk 12]
```diff
@@ -0,0 +1,8 @@
+## Fine-tuning Wav2Vec2
+
+The `run_training.py` script allows one to finetune pretrained Wav2Vec2 models that can be found [here](https://huggingface.co/models?search=facebook/wav2vec2).
```
Comment (sgugger) [inline examples/research_projects/wav2vec2/README.md:3]: We usually name the run script with the ask, so `run_wav2vec`?

File: examples/research_projects/wav2vec2/run_training.py (hunk 5 above)
Comment (sgugger) [reply examples/research_projects/wav2vec2/run_training.py:1]: I 100%-agree

File: src/transformers/models/wav2vec2/processing_wav2vec2.py [hunk 13]
```diff
@@ -115,6 +115,16 @@ def __call__(self, *args, **kwargs):
         """
         return self.current_processor(*args, **kwargs)
 
+    def pad(self, *args, **kwargs):
+        """
+        When used in normal mode, this method forwards all its arguments to Wav2Vec2FeatureExtractor's
+        :meth:`~transformers.Wav2Vec2FeatureExtractor.pad` and returns its output. If used in the context
+        :meth:`~transformers.Wav2Vec2Processor.as_target_processor` this method forwards all its arguments to
+        Wav2Vec2CTCTokenizer's :meth:`~transformers.Wav2Vec2CTCTokenizer.pad`. Please refer to the doctsring of the
```
Comment (LysandreJik) [inline src/transformers/models/wav2vec2/processing_wav2vec2.py:123]:
```suggestion
        Wav2Vec2CTCTokenizer's :meth:`~transformers.Wav2Vec2CTCTokenizer.pad`. Please refer to the docstring of the
```

File: examples/research_projects/wav2vec2/run_training.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    from apex import amp
+
+
+if version.parse(torch.__version__) >= version.parse("1.6"):
+    _is_native_amp_available = True
+    from torch.cuda.amp import autocast
+
+
+model = Wav2Vec2ForCTC.from_pretrained("facebook/wav2vec2-base", gradient_checkpointing=True)
+processor = Wav2Vec2Processor.from_pretrained("facebook/wav2vec2-base")
```
Comment (sgugger) [inline examples/research_projects/wav2vec2/run_training.py:26]: The README said it was for any wav2vec checkpoint, so make `""facebook/wav2vec2-base"` an argument (the ususal model_name_or_path)?

File: examples/research_projects/wav2vec2/run_training.py [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+            self.scaler.scale(loss).backward()
+        elif self.use_apex:
+            with amp.scale_loss(loss, self.optimizer) as scaled_loss:
+                scaled_loss.backward()
+        elif self.deepspeed:
+            self.deepspeed.backward(loss)
+        else:
+            loss.backward()
+
+        return loss.detach()
```
Comment (sgugger) [inline examples/research_projects/wav2vec2/run_training.py:190]: Cool example of a subclass!

File: examples/research_projects/wav2vec2/run_training.py [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+    save_total_limit=3,
+    save_steps=500,
+    eval_steps=100,
+    logging_steps=50,
+    learning_rate=5e-4,
+    warmup_steps=3000,
+    group_by_length=True,
+    logging_dir="./logs",
+    fp16=True,
+)
```
Comment (sgugger) [inline examples/research_projects/wav2vec2/run_training.py:208]: Why not make those parametrizable as usual? And then the default can be set in a `.sh` file?

File: tests/test_modeling_wav2vec2.py [hunk 17]
```diff
@@ -155,6 +154,60 @@ def create_and_check_batch_inference(self, config, input_values, *args):
             batch_output = batch_outputs[i : i + 1, : output.shape[1]]
             self.parent.assertTrue(torch.allclose(output, batch_output, atol=1e-3))
 
+    def check_ctc_loss(self, config, input_values, *args):
```
Comment (LysandreJik) [inline tests/test_modeling_wav2vec2.py:157]: Very nice test!

File: src/transformers/models/wav2vec2/configuration_wav2vec2.py [hunk 18]
```diff
@@ -92,6 +92,33 @@ class Wav2Vec2Config(PretrainedConfig):
             Whether do apply `stable` layer norm architecture of the Transformer encoder. ``do_stable_layer_norm is
             True`` corresponds to applying layer norm before the attention layer, whereas ``do_stable_layer_norm is
             False`` corresponds to applying layer norm after the attention layer.
+        freeze_feat_extract_train (:obj:`bool`, `optional`, defaults to :obj:`True`):
+            Whether to freeze the weights of the feature extractor when training.
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/configuration_wav2vec2.py:96]: Not sure if this one should be in the model config or a specific training argument. No strong opinion if you prefer it there though.

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 19]
```diff
@@ -178,6 +243,11 @@ def __init__(self, config):
             )
         self.conv_layers = nn.ModuleList(conv_layers)
 
+        if config.freeze_feat_extract_train:
+            # freeze feature_extractor
+            for param in self.parameters():
+                param.requires_grad = False
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:249]: I would put this in a model method personally (and the config arg as a training argument, the script would then be responsible to call the method on the model).

The problem with putting it in the config is that it will be saved with the pretrained model which will then do this forever and ever (unless someone goes out of their way to remove it) whereas it seems just a training artifact to me. It's a bit as if we stored the learning rate in the model config and used it for all subsequents fine-tunings.

**Review by LysandreJik — APPROVED** [review]
Very cool! The training code is clear and simple, it's nice.

I have not deeply reviewed the Trainer changes but from a quick look it looks good. The rest is great, very cool tests.

**Review by sgugger — APPROVED** [review]
Very nice! I just have one main comment (see below). Also, don't forget to add a requirements.txt in your research project folder!

File: examples/research_projects/wav2vec2/README.md (hunk 12 above)
Comment (patil-suraj) [reply examples/research_projects/wav2vec2/README.md:3]: I would prefer `run_asr`, we name scripts for tasks, `wav2vec` is not a task

File: examples/research_projects/wav2vec2/run_training.py (hunk 10 above)
Comment (patil-suraj) [reply examples/research_projects/wav2vec2/run_training.py:12]: think We should add this metric to `datasets` now, as we have two speech models and many speech datasets coming to the datasets lib

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 8 above)
Comment (patil-suraj) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:1063]: I agree the first option. We don't replace -100 with pad for any model, so I think the first option is good.

**Review by patil-suraj — APPROVED** [review]
Looks good to me! Very cool!

File: src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to_pytorch.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to_pytorch.py:32]: No it was only for new models

**Comment by YUYUNHUANG** [comment]
lifesaver!

File: examples/research_projects/wav2vec2/run_training.py (hunk 10 above)
Comment (patrickvonplaten) [reply examples/research_projects/wav2vec2/run_training.py:12]: Actually it's already in the dataset, will use it instead

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 19 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:249]: Hmmm, I don't fully agree here since the default case is that the feature extraction layers are not fine-tuned anymore after pertaining, but I definitely see your point! So good for me to have a model method instead and call it from the fine-tuning script!

Comment (sgugger) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py]: > Also, we have configuration arguments like gradient_checkpointing (do you think they should not be in the config either?)

Yes, absolutely. I've discussed it with @LysandreJik and will make a PR when I have a bit of time to work on this. For the same reason I think `gradient_chekpointing` should be a model method that we activate in the training scripts when the right training arg is passed (of course in the case of gradient checkpointing, there will be a deprecation cycle and proper BC since it's been around for a while).
